runtime/registry.go

```diff
@@ -540,6 +545,48 @@ func (r *registryCache) emitHeartbeatForInstance(inst *drivers.Instance) {
 	)
 }
 
+// updateProjectConfig updates the project config for the given instance.
+// This is essentially a copy of ProjectParserReconciler's reconcileProjectConfig and is done before starting the controller
+// to ensure that when controller first starts, it doesn’t immediately restart due to changed variables
+func (r *registryCache) updateProjectConfig(iwc *instanceWithController) error {
+	repo, release, err := r.rt.Repo(iwc.ctx, iwc.instanceID)
+	if err != nil {
+		return err
+	}
+	defer release()
+
+	inst, err := r.rt.Instance(iwc.ctx, iwc.instanceID)
+	if err != nil {
+		return err
+	}
+
+	rillYAML, err := rillv1.ParseRillYAML(iwc.ctx, repo, iwc.instanceID)
+	if err != nil {
+		return err
+	}
+	inst.ProjectOLAPConnector = rillYAML.OLAPConnector
+	conns := make([]*runtimev1.Connector, 0, len(rillYAML.Connectors))
+	for _, c := range rillYAML.Connectors {
+		conns = append(conns, &runtimev1.Connector{
+			Type:   c.Type,
+			Name:   c.Name,
+			Config: c.Defaults,
+		})
+	}
+	inst.ProjectConnectors = conns
+	inst.ProjectVariables = make(map[string]string)
+	for _, v := range rillYAML.Variables {
+		inst.ProjectVariables[v.Name] = v.Default
+	}
+
+	dotEnv, err := rillv1.ParseDotEnv(iwc.ctx, repo, iwc.instanceID)
+	if err != nil {
+		return err
+	}
+	maps.Copy(inst.ProjectVariables, dotEnv)
+	return r.rt.EditInstance(iwc.ctx, inst, false)
```

Comment:
Would it make sense to we extract this part into a util function directly on the `Runtime` object so it can be called from both the `ProjectParserReconciler` and here? Like `rt.UpdateInstanceWithRillYAML(ctx, instanceID, rillYAML, dotEnv)`.

It's probably best to centralize it one place because this is actually already missing handling of a property added a few hours back: https://github.com/rilldata/rill/blob/d0cba69b5d7ec3f08ca914666cfbae16d13ed0a4/runtime/reconcilers/project_parser.go#L358